Can't build stripped down S3 flavor of cachepot

I want a no-frills cachepot for my S3-powered CI environment but it won't build. My environment:

    % rustc --version
    rustc 1.61.0-nightly (52b34550a 2022-03-15)
    % cargo --version
    cargo 1.61.0-nightly (65c826642 2022-03-09)

The error:

    % cargo install cachepot --version 0.1.0-rc.1 --no-default-features --features=s3
    [[[ SNIP ]]]
    error[E0432]: unresolved import `crate::config::WorkerUrl`
      --> /Users/xlange/.cargo/registry/src/github.com-1ecc6299db9ec823/cachepot-0.1.0-rc.1/src/dist/mod.rs:16:5
       |
    16 | use crate::config::WorkerUrl;
       |     ^^^^^^^^^^^^^^^^^^^^^^^^ no `WorkerUrl` in `config`
    
    error[E0412]: cannot find type `WorkerUrl` in module `crate::config`
       --> /Users/xlange/.cargo/registry/src/github.com-1ecc6299db9ec823/cachepot-0.1.0-rc.1/src/compiler/compiler.rs:725:23
        |
    725 |     Ok(crate::config::WorkerUrl),
        |                       ^^^^^^^^^ not found in `crate::config`
    
    Some errors have detailed explanations: E0412, E0432.
    For more information about an error, try `rustc --explain E0412`.
    error: could not compile `cachepot` due to 2 previous errors
    warning: build failed, waiting for other jobs to finish...
    error: failed to compile `cachepot v0.1.0-rc.1`, intermediate artifacts can be found at `/var/folders/mc/3pj1vp8s239fn610rbv56f440000gp/T/cargo-installTYYXXo`
    
    Caused by:
      build failed

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -86,7 +86,7 @@ tokio-util = { version = "0.6", features = ["codec"] }
 tower = "0.4"
 toml = "0.5"
 untrusted = { version = "0.7", optional = true }
-url = { version = "2", optional = true }
+url = "2"
 uuid = { version = "0.8", features = ["v4"] }
 walkdir = "2"
 # by default which pulls in an outdated failure version
@@ -137,15 +137,15 @@ features = [
 [features]
 default = ["all"]
 all = ["dist-client", "redis", "s3", "memcached", "gcs", "azure"]
-azure = ["chrono", "hyper", "hyperx", "url", "hmac", "md-5", "sha2"]
+azure = ["chrono", "hyper", "hyperx", "hmac", "md-5", "sha2"]
 s3 = ["chrono", "hyper", "hyper-tls", "hyperx", "reqwest", "rusoto_core", "rusoto_s3", "hmac", "sha-1"]
-gcs = ["chrono", "hyper", "hyperx", "percent-encoding", "reqwest", "ring", "untrusted", "url"]
+gcs = ["chrono", "hyper", "hyperx", "percent-encoding", "reqwest", "ring", "untrusted"]
 memcached = ["memcached-rs"]
 native-zlib = []
 # Enable features that require unstable features of Nightly Rust.
 unstable = []
 # Enables distributed support in the cachepot client
-dist-client = ["ar", "flate2", "hyper", "hyperx", "reqwest/stream", "url", "sha2", "tokio/fs"]
+dist-client = ["ar", "flate2", "hyper", "hyperx", "reqwest/stream", "sha2", "tokio/fs"]
 # Enables the cachepot-dist binary
 dist-worker = ["chrono", "crossbeam-utils", "jsonwebtoken", "flate2", "hyperx", "libmount", "nix", "reqwest",  "sha2", "syslog", "void", "version-compare", "warp", "thiserror"]
 # Enables dist tests with external requirements

--- a/src/compiler/rust.rs
+++ b/src/compiler/rust.rs
@@ -157,16 +157,20 @@ pub struct RustCompilation {
     sysroot: PathBuf,
     /// All arguments passed to rustc
     arguments: Vec<Argument<ArgData>>,
+    #[cfg_attr(not(feature = "dist-client"), allow(dead_code))]
     /// The compiler inputs.
     inputs: Vec<PathBuf>,
     /// The compiler outputs.
     outputs: HashMap<String, PathBuf>,
+    #[cfg_attr(not(feature = "dist-client"), allow(dead_code))]
     /// The directories searched for rlibs
     crate_link_paths: Vec<PathBuf>,
     /// The crate name being compiled.
     crate_name: String,
+    #[cfg_attr(not(feature = "dist-client"), allow(dead_code))]
     /// The crate types that will be generated
     crate_types: CrateTypes,
+    #[cfg_attr(not(feature = "dist-client"), allow(dead_code))]
     /// If dependency info is being emitted, the name of the dep info file.
     dep_info: Option<PathBuf>,
     /// The current working directory

--- a/src/config.rs
+++ b/src/config.rs
@@ -16,7 +16,6 @@ use crate::util::fs::{self, File};
 use directories::ProjectDirs;
 use regex::Regex;
 use serde::de::{Deserialize, DeserializeOwned, Deserializer};
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
 use serde::ser::{Serialize, Serializer};
 use std::collections::HashMap;
 use std::env;
@@ -89,10 +88,9 @@ pub fn parse_size(val: &str) -> Option<u64> {
         })
 }
 
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub struct HTTPUrl(reqwest::Url);
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 impl Serialize for HTTPUrl {
     fn serialize<S>(&self, serializer: S) -> StdResult<S::Ok, S::Error>
     where
@@ -101,7 +99,7 @@ impl Serialize for HTTPUrl {
         serializer.serialize_str(self.0.as_str())
     }
 }
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 impl<'a> Deserialize<'a> for HTTPUrl {
     fn deserialize<D>(deserializer: D) -> StdResult<Self, D::Error>
     where
@@ -113,7 +111,7 @@ impl<'a> Deserialize<'a> for HTTPUrl {
         Ok(HTTPUrl(url))
     }
 }
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 fn parse_http_url(url: &str) -> Result<reqwest::Url> {
     let url = reqwest::Url::parse(url)?;
 
@@ -126,7 +124,7 @@ fn parse_http_url(url: &str) -> Result<reqwest::Url> {
     }
     Ok(url)
 }
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 impl HTTPUrl {
     pub fn from_url(u: reqwest::Url) -> Self {
         HTTPUrl(u)
@@ -143,7 +141,7 @@ impl HTTPUrl {
             .expect("HTTPUrl always has a valid host; qed")
     }
 }
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 impl FromStr for HTTPUrl {
     type Err = anyhow::Error;
 
@@ -156,12 +154,13 @@ impl FromStr for HTTPUrl {
         Ok(HTTPUrl(url))
     }
 }
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 impl fmt::Display for HTTPUrl {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(f, "{}", self.0)
     }
 }
+
 // This is the extended `HTTPUrl` with re-implemented
 // `Serialize` and `Deserialize` traits. It is known that
 // servers only use the `https` protocol and as such we can
@@ -169,10 +168,9 @@ impl fmt::Display for HTTPUrl {
 // of such format is that it can be used in un-urlencoded form as params:
 //
 // `https://localhost:10500/api/v1/scheduler/server_certificate/localhost:10603/`
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub struct WorkerUrl(pub HTTPUrl);
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 impl Serialize for WorkerUrl {
     fn serialize<S>(&self, serializer: S) -> StdResult<S::Ok, S::Error>
     where
@@ -187,7 +185,7 @@ impl Serialize for WorkerUrl {
         serializer.serialize_str(&helper)
     }
 }
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 impl<'a> Deserialize<'a> for WorkerUrl {
     fn deserialize<D>(deserializer: D) -> StdResult<Self, D::Error>
     where
@@ -200,7 +198,7 @@ impl<'a> Deserialize<'a> for WorkerUrl {
         Ok(WorkerUrl(HTTPUrl(url)))
     }
 }
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 impl FromStr for WorkerUrl {
     type Err = anyhow::Error;
 
@@ -210,7 +208,7 @@ impl FromStr for WorkerUrl {
         Ok(WorkerUrl(HTTPUrl::from_str(&helper)?))
     }
 }
-#[cfg(any(feature = "dist-client", feature = "dist-worker"))]
+
 impl fmt::Display for WorkerUrl {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(

